### DISCUSSION TITLE
Adding constant momentum source in cylindrical coordinates

### DISCUSSION THREAD

**Discussion by rahulgaur104** [PrincetonUniversity/athena#432] [Q&A]
Hi everyone,

I want to add a constant force/volume(force density?) source in the azimuthal direction to the cylindrical system(r, phi, z) that I am simulating. The snippet of my code that does this is shown below:

```
//----------------------------------------------------------------------------------------
//! User-defined azimuthal momentum source term throughout the plasma
//----------------------------------------------------------------------------------------

void ExtSource(MeshBlock *pmb, const Real time, const Real dt,
              const AthenaArray<Real> &prim, const AthenaArray<Real> &prim_scalar,
              const AthenaArray<Real> &bcc, AthenaArray<Real> &cons,
              AthenaArray<Real> &cons_scalar) {


  for (int k=pmb->ks; k<=pmb->ke; ++k) {
    for (int j=pmb->js; j<=pmb->je; ++j) {
      for (int i=pmb->is; i<=pmb->ie; ++i) {
         cons(IM2,k,j,i) += SourceStrength*110000*dt*(1/(1 + std::exp(-60*(time-0.01)))) ;
        }
      }
    }
  return;
}
```

where SourceStrength is a number that is provided in the athinput file. Is this the right way to add a force/volume source in cylindrical coordinates that is independent of x1? I was looking at Skinner and Ostriker's paper ([here](https://arxiv.org/pdf/1004.2487.pdf)) and I saw the source term in the momentum equation that contained factors of x1(or r).

To understand this, I looked at line 320 of /src/coordinates/cylindrical.cpp which confused me even further. Is `coord_src2_i_(i)` the same as the constant momentum density source term `SourceStrength`?

My ultimate objective is to add a constant force/volume source in the azimuthal(x2(or phi)) direction. Any feedback would be appreciated.

Thank you,
Rahul

**Comment by c-white** [comment]
The code you wrote should be correct.

The paper you cite (see also the equations near 2.14 of [this PhD thesis](https://dataspace.princeton.edu/handle/88435/dsp017s75df84b), for more details) is adding a phi-momentum source term, but in a different context. The conservation laws/evolution equations have the form

&part;<sub>t</sub>(conserved) + div(flux) = S<sub>physical</sub>(primitive).

In non-Cartesian coordinates, the divergence operator has extra/modified terms, so this becomes

&part;<sub>t</sub>(conserved) + &part;<sub>x1</sub>(flux1) + &part;<sub>x2</sub>(flux2) + &part;<sub>x3</sub>(flux3) = S<sub>geometric</sub>(primitive) + S<sub>physical</sub>(primitive).

The problem is, if you naively integrate this equation over a single cell to turn it into a finite-volume scheme, the fluxes evaluated at cell faces won't combine with the cell-averaged geometric source term to exactly get the discrete divergence operator you want, causing things like secular errors in angular momentum. As it turns out, you can write the geometric source term in terms of face-centered fluxes (rather than cell-centered conserved or primitive variables):

&part;<sub>t</sub>(conserved) + &part;<sub>x1</sub>(flux1) + &part;<sub>x2</sub>(flux2) + &part;<sub>x3</sub>(flux3) = S'<sub>geometric</sub>(fluxes) + S<sub>physical</sub>(primitive).

Even if S<sub>geometric</sub> doesn't depend on position, S'<sub>geometric</sub> may well do so.

However, none of this affects the physical source terms you are likely to be implementing as a user.

**Reply by rahulgaur104** [reply]
Thanks, Chris! If I understand you correctly, the user-implemented term SourceStrength*dt, SourceStrength being a constant, has dimensions of Force/volume and is independent of all coordinates.

**Reply by c-white** [reply]
Yes, SourceStrength should be force/volume.
